File: apps/multiplatform/common/src/commonMain/kotlin/chat/simplex/common/views/chat/group/GroupMentions.kt
package chat.simplex.common.views.chat.group

import androidx.compose.animation.core.*
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.*
import androidx.compose.material.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.text.TextRange
import androidx.compose.ui.unit.*
import chat.simplex.common.model.*
import chat.simplex.common.platform.*
import chat.simplex.common.ui.theme.*
import chat.simplex.common.views.chat.*
import chat.simplex.common.views.chatlist.setGroupMembers
import chat.simplex.common.views.helpers.*
import chat.simplex.res.MR
import kotlinx.coroutines.launch
import kotlin.text.CharCategory.*

val punctuation = setOf(
  DASH_PUNCTUATION, START_PUNCTUATION, END_PUNCTUATION,
  CONNECTOR_PUNCTUATION, OTHER_PUNCTUATION
)

private val PICKER_ROW_SIZE = MEMBER_ROW_AVATAR_SIZE + (MEMBER_ROW_VERTICAL_PADDING * 2f)
private val MAX_PICKER_HEIGHT = (PICKER_ROW_SIZE * 4) + (MEMBER_ROW_AVATAR_SIZE + MEMBER_ROW_VERTICAL_PADDING - 4.dp)

@Composable
fun GroupMentions(
  rhId: Long?,
  composeState: MutableState<ComposeState>,
  composeViewFocusRequester: FocusRequester?,
  chatInfo: ChatInfo.Group
) {
  val maxHeightInPx = with(LocalDensity.current) { windowHeight().toPx() }
  val isVisible = remember { mutableStateOf(false) }
  val offsetY = remember { Animatable(maxHeightInPx) }

  val currentMessage = remember { mutableStateOf(composeState.value.message) }
  val mentionName = remember { mutableStateOf("") }
  val mentionRange = remember { mutableStateOf<TextRange?>(null) }
  val mentionMemberId = remember { mutableStateOf<String?>(null) }
  val filteredMembers = remember {
    derivedStateOf {
      val members = chatModel.groupMembers.value
        .filter {
          val status = it.memberStatus
          status != GroupMemberStatus.MemLeft && status != GroupMemberStatus.MemRemoved && status != GroupMemberStatus.MemInvited
        }
        .sortedByDescending { it.memberRole }

      if (mentionName.value.isEmpty()) {
        members
      } else {
        members.filter { it.memberProfile.anyNameContains(mentionName.value) }
      }
    }
  }
  val scope = rememberCoroutineScope()

  suspend fun closeMembersPicker() {
    isVisible.value = false
    if (offsetY.value != 0f) {
      return
    }

    offsetY.animateTo(
      targetValue = maxHeightInPx,
      animationSpec = mentionPickerAnimSpec()
    )
    mentionName.value = ""
    mentionRange.value = null
    mentionMemberId.value = null
  }

  fun messageChanged(msg: ComposeMessage, parsedMsg: List<FormattedText>) {
    removeUnusedMentions(composeState, parsedMsg)
    val selected = selectedMarkdown(parsedMsg, msg.selection)

    if (selected != null) {
      val (ft, r) = selected

      when (ft.format) {
        is Format.Mention -> {
          isVisible.value = true
          mentionName.value = ft.format.memberName
          mentionRange.value = r
          mentionMemberId.value = composeState.value.mentions[mentionName.value]?.memberId
          if (!chatModel.membersLoaded.value) {
            scope.launch {
              setGroupMembers(rhId, chatInfo.groupInfo, chatModel)
            }
          }
          return
        }
        null -> {
          val pos = msg.selection.start
          if (msg.selection.length == 0 && getCharacter(msg.text, pos - 1)?.first == "@") {
            val prevChar = getCharacter(msg.text, pos - 2)?.first
            if (prevChar == null || prevChar == " " || prevChar == "\n") {
              isVisible.value = true
              mentionName.value = ""
              mentionRange.value = TextRange(pos - 1, pos)
              mentionMemberId.value = null
              scope.launch {
                setGroupMembers(rhId, chatInfo.groupInfo, chatModel)
              }
              return
            }
          }
        }
        else -> {}
      }
    }
    scope.launch {
      closeMembersPicker()
    }
  }

  fun addMemberMention(member: GroupMember, range: TextRange) {
    val mentions = composeState.value.mentions.toMutableMap()
    val existingMention = mentions.entries.firstOrNull {
      it.value.memberId == member.memberId
    }
    val newName = existingMention?.key ?: composeState.value.mentionMemberName(member.memberProfile.displayName)
    mentions[newName] = CIMention(member)
    var msgMention = if (newName.contains(" ") || (newName.lastOrNull()?.category in punctuation))
                      "@'$newName'"
                      else "@$newName"
    var newPos = range.start + msgMention.length
    val newMsgLength = composeState.value.message.text.length + msgMention.length - range.length
    if (newPos == newMsgLength) {
      msgMention += " "
      newPos += 1
    }

    val msg = composeState.value.message.text.replaceRange(
      range.start,
      range.end,
      msgMention
    )
    composeState.value = composeState.value.copy(
      message = ComposeMessage(msg, TextRange(newPos)),
      parsedMessage = parseToMarkdown(msg) ?: FormattedText.plain(msg),
      mentions = mentions
    )

    composeViewFocusRequester?.requestFocus()

    scope.launch {
      closeMembersPicker()
    }
  }

  LaunchedEffect(composeState.value.parsedMessage) {
    currentMessage.value = composeState.value.message
    messageChanged(currentMessage.value, composeState.value.parsedMessage)
  }

//  KeyChangeEffect(composeState.value.message.selection) {
//    // This condition is needed to prevent messageChanged called twice,
//    // because composeState.formattedText triggers later when message changes.
//    // The condition is only true if position changed without text change
//    if (currentMessage.value.text == composeState.value.message.text) {
//      messageChanged(currentMessage.value, composeState.value.parsedMessage)
//    }
//  }

  LaunchedEffect(isVisible.value) {
    if (isVisible.value) {
      offsetY.animateTo(
        targetValue = 0f,
        animationSpec = mentionPickerAnimSpec()
      )
    }
  }
  Box(
    modifier = Modifier
      .fillMaxSize()
      .offset { IntOffset(0, offsetY.value.toInt()) }
      .clickable(indication = null, interactionSource = remember { MutableInteractionSource() }) {
        scope.launch { closeMembersPicker() }
      },
    contentAlignment = Alignment.BottomStart
  ) {
    val showMaxReachedBox = composeState.value.mentions.size >= MAX_NUMBER_OF_MENTIONS && isVisible.value && composeState.value.mentions[mentionName.value] == null
    LazyColumnWithScrollBarNoAppBar(
      Modifier
        .heightIn(max = MAX_PICKER_HEIGHT)
        .background(MaterialTheme.colors.surface),
      maxHeight = remember { mutableStateOf(MAX_PICKER_HEIGHT) },
      containerAlignment = Alignment.BottomEnd
    ) {
      if (showMaxReachedBox) {
        stickyHeader {
          MaxMentionsReached()
        }
      }
      itemsIndexed(filteredMembers.value, key = { _, item -> item.groupMemberId }) { i, member ->
        if (i != 0 || !showMaxReachedBox) {
          Divider()
        }
        val mentioned = mentionMemberId.value == member.memberId
        val disabled = composeState.value.mentions.size >= MAX_NUMBER_OF_MENTIONS && !mentioned
        Row(
          Modifier
            .fillMaxWidth()
            .alpha(if (disabled) 0.6f else 1f)
            .clickable(enabled = !disabled) {
              val range = mentionRange.value ?: return@clickable
              val mentionMemberValue = mentionMemberId.value

              if (mentionMemberValue != null) {
                if (mentionMemberValue != member.memberId) {
                  addMemberMention(member, range)
                } else {
                  return@clickable
                }
              } else {
                addMemberMention(member, range)
              }
            }
            .padding(horizontal = DEFAULT_PADDING_HALF),
          verticalAlignment = Alignment.CenterVertically
        ) {
          MemberRow(
            member,
            infoPage = false,
            showlocalAliasAndFullName = true,
            selected = mentioned
          )
        }
      }
    }
  }
}

@Composable
private fun MaxMentionsReached() {
  Column(Modifier.background(MaterialTheme.colors.surface)) {
    Divider()
    Row(
      Modifier.fillMaxWidth(),
      verticalAlignment = Alignment.CenterVertically,
    ) {
      Text(
        String.format(generalGetString(MR.strings.max_group_mentions_per_message_reached), MAX_NUMBER_OF_MENTIONS),
        Modifier.padding(12.dp),
      )
    }
    Divider()
  }
}

private fun getCharacter(s: String, pos: Int): Pair<CharSequence, IntRange>? {
  return if (pos in s.indices) {
    val char = s.subSequence(pos, pos + 1)
    char to (pos until pos + 1)
  } else {
    null
  }
}

private fun selectedMarkdown(
  parsedMsg: List<FormattedText>,
  range: TextRange
): Pair<FormattedText, TextRange>? {
  if (parsedMsg.isEmpty()) return null

  var i = 0
  var pos = 0

  while (i < parsedMsg.size && pos + parsedMsg[i].text.length < range.start) {
    pos += parsedMsg[i].text.length
    i++
  }

  return if (i >= parsedMsg.size || range.end > pos + parsedMsg[i].text.length) {
    null
  } else {
    parsedMsg[i] to TextRange(pos, pos + parsedMsg[i].text.length)
  }
}

private fun removeUnusedMentions(composeState: MutableState<ComposeState>, parsedMsg: List<FormattedText>) {
  val usedMentions = parsedMsg.mapNotNull { ft ->
    when (ft.format) {
      is Format.Mention -> ft.format.memberName
      else -> null
    }
  }.toSet()

  if (usedMentions.size < composeState.value.mentions.size) {
    composeState.value = composeState.value.copy(
      mentions = composeState.value.mentions.filterKeys { it in usedMentions }
    )
  }
}

File: apps/ios/Shared/Views/Chat/Group/GroupMentions.swift
//
//  GroupMentions.swift
//  SimpleX (iOS)
//
//  Created by Diogo Cunha on 30/01/2025.
//  Copyright © 2025 SimpleX Chat. All rights reserved.
//

import SwiftUI
import SimpleXChat

let MENTION_START: Character = "@"
let QUOTE: Character = "'"
let MEMBER_ROW_SIZE: CGFloat = 60
let MAX_VISIBLE_MEMBER_ROWS: CGFloat = 4.8

struct GroupMentionsView: View {
    @EnvironmentObject var m: ChatModel
    @EnvironmentObject var theme: AppTheme
    var groupInfo: GroupInfo
    @Binding var composeState: ComposeState
    @Binding var selectedRange: NSRange
    @Binding var keyboardVisible: Bool

    @State private var isVisible = false
    @State private var currentMessage: String = ""
    @State private var mentionName: String = ""
    @State private var mentionRange: NSRange?
    @State private var mentionMemberId: String?
    @State private var sortedMembers: [GMember] = []

    var body: some View {
        ZStack(alignment: .bottom) {
            if isVisible {
                let filtered = filteredMembers()
                if filtered.count > 0 {
                    Color.white.opacity(0.01)
                        .edgesIgnoringSafeArea(.all)
                        .onTapGesture {
                            isVisible = false
                        }
                    VStack(spacing: 0) {
                        Spacer()
                        Divider()
                        let scroll = ScrollView {
                            LazyVStack(spacing: 0) {
                                ForEach(Array(filtered.enumerated()), id: \.element.wrapped.groupMemberId) { index, member in
                                    let mentioned = mentionMemberId == member.wrapped.memberId
                                    let disabled = composeState.mentions.count >= MAX_NUMBER_OF_MENTIONS && !mentioned
                                    ZStack(alignment: .bottom) {
                                        memberRowView(member.wrapped, mentioned)
                                            .contentShape(Rectangle())
                                            .disabled(disabled)
                                            .opacity(disabled ? 0.6 : 1)
                                            .onTapGesture {
                                                memberSelected(member)
                                            }
                                            .padding(.horizontal)
                                            .frame(height: MEMBER_ROW_SIZE)

                                        Divider()
                                            .padding(.leading)
                                            .padding(.leading, 48)
                                    }
                                }
                            }
                        }
                            .frame(maxHeight: MEMBER_ROW_SIZE * min(MAX_VISIBLE_MEMBER_ROWS, CGFloat(filtered.count)))
                            .background(Color(UIColor.systemBackground))

                        if #available(iOS 16.0, *) {
                            scroll.scrollDismissesKeyboard(.never)
                        } else {
                            scroll
                        }
                    }
                }
            }
        }
        .onChange(of: composeState.parsedMessage) { parsedMsg in
            currentMessage = composeState.message
            messageChanged(currentMessage, parsedMsg, selectedRange)
        }
        .onChange(of: selectedRange) { r in
            // This condition is needed to prevent messageChanged called twice,
            // because composeState.formattedText triggers later when message changes.
            // The condition is only true if position changed without text change
            if currentMessage == composeState.message {
                messageChanged(currentMessage, composeState.parsedMessage, r)
            }
        }
        .onAppear {
            currentMessage = composeState.message
        }
    }
    
    private func filteredMembers() -> [GMember] {
        let s = mentionName.lowercased()
        return s.isEmpty
        ? sortedMembers
        : sortedMembers.filter { $0.wrapped.localAliasAndFullName.localizedLowercase.contains(s) }
    }

    private func messageChanged(_ msg: String, _ parsedMsg: [FormattedText], _ range: NSRange) {
        removeUnusedMentions(parsedMsg)
        if let (ft, r) = selectedMarkdown(parsedMsg, range) {
            switch ft.format {
            case let .mention(name):
                isVisible = true
                mentionName = name
                mentionRange = r
                mentionMemberId = composeState.mentions[name]?.memberId
                if !m.membersLoaded {
                    Task {
                        await m.loadGroupMembers(groupInfo)
                        sortMembers()
                    }
                }
                return
            case .none: () //
                let pos = range.location
                if range.length == 0, let (at, atRange) = getCharacter(msg, pos - 1), at == "@" {
                    let prevChar = getCharacter(msg, pos - 2)?.char
                    if prevChar == nil || prevChar == " " || prevChar == "\n" {
                        isVisible = true
                        mentionName = ""
                        mentionRange = atRange
                        mentionMemberId = nil
                        Task {
                            await m.loadGroupMembers(groupInfo)
                            sortMembers()
                        }
                        return
                    }
                }
            default: ()
            }
        }
        closeMemberList()
    }

    private func sortMembers() {
        sortedMembers = m.groupMembers.filter({ m in
            let status = m.wrapped.memberStatus
            return status != .memLeft && status != .memRemoved && status != .memInvited
        })
        .sorted { $0.wrapped.memberRole > $1.wrapped.memberRole }
    }

    private func removeUnusedMentions(_ parsedMsg: [FormattedText]) {
        let usedMentions: Set<String> = Set(parsedMsg.compactMap { ft in
            if case let .mention(name) = ft.format { name } else { nil }
        })
        if usedMentions.count < composeState.mentions.count  {
            composeState = composeState.copy(mentions: composeState.mentions.filter({ usedMentions.contains($0.key) }))
        }
    }

    private func getCharacter(_ s: String, _ pos: Int) -> (char: String.SubSequence, range: NSRange)? {
        if pos < 0 || pos >= s.count { return nil }
        let r = NSRange(location: pos, length: 1)
        return if let range = Range(r, in: s) {
            (s[range], r)
        } else {
            nil
        }
    }
    
    private func selectedMarkdown(_ parsedMsg: [FormattedText], _ range: NSRange) -> (FormattedText, NSRange)? {
        if parsedMsg.isEmpty { return nil }
        var i = 0
        var pos: Int = 0
        while i < parsedMsg.count && pos + parsedMsg[i].text.count < range.location {
            pos += parsedMsg[i].text.count
            i += 1
        }
        // the second condition will be true when two markdowns are selected
        return i >= parsedMsg.count || range.location + range.length > pos + parsedMsg[i].text.count
            ? nil
            : (parsedMsg[i], NSRange(location: pos, length: parsedMsg[i].text.count))
    }
    
    private func memberSelected(_ member: GMember) {
        if let range = mentionRange, mentionMemberId == nil || mentionMemberId != member.wrapped.memberId {
            addMemberMention(member, range)
        }
    }
    
    private func addMemberMention(_ member: GMember, _ r: NSRange) {
        guard let range = Range(r, in: composeState.message) else { return }
        var mentions = composeState.mentions
        var newName: String
        if let mm = mentions.first(where: { $0.value.memberId == member.wrapped.memberId }) {
            newName = mm.key
        } else {
            newName = composeState.mentionMemberName(member.wrapped.memberProfile.displayName)
        }
        mentions[newName] = CIMention(groupMember: member.wrapped)
        var msgMention = newName.contains(" ") || newName.last?.isPunctuation == true
                        ? "@'\(newName)'"
                        : "@\(newName)"
        var newPos = r.location + msgMention.count
        let newMsgLength = composeState.message.count + msgMention.count - r.length
        print(newPos)
        print(newMsgLength)
        if newPos == newMsgLength {
            msgMention += " "
            newPos += 1
        }
        composeState = composeState.copy(
            message: composeState.message.replacingCharacters(in: range, with: msgMention),
            mentions: mentions
        )
        selectedRange = NSRange(location: newPos, length: 0)
        closeMemberList()
        keyboardVisible = true
    }
    
    private func closeMemberList() {
        isVisible = false
        mentionName = ""
        mentionRange = nil
        mentionMemberId = nil
    }
    
    private func memberRowView(_ member: GroupMember, _ mentioned: Bool) -> some View {
        return HStack{
            MemberProfileImage(member, size: 38)
                .padding(.trailing, 2)
            VStack(alignment: .leading) {
                let t = Text(member.localAliasAndFullName).foregroundColor(member.memberIncognito ? .indigo : theme.colors.onBackground)
                (member.verified ? memberVerifiedShield() + t : t)
                    .lineLimit(1)
            }
            Spacer()
            if mentioned {
                Image(systemName: "checkmark")
            }
        }

        func memberVerifiedShield() -> Text {
            (Text(Image(systemName: "checkmark.shield")) + textSpace)
                .font(.caption)
                .baselineOffset(2)
                .kerning(-2)
                .foregroundColor(theme.colors.secondary)
        }
    }
}
